### PR TITLE
Book: First draft of signals chapter

### LINKED ISSUE #50: Book: First draft of Signals Chapter
Add a first draft for the signal handling chapter of the CLI book (#45).

- [ ] What are signals
- [ ] Easy crate to deal with Ctrl+C
- [ ] Handling signals with threads and channels?

Here's something to get you started:

> Processes like command line applications need to react to signals sent by the operating system.
> The most common example is probably <kbd>Ctrl</kbd>+<kbd>C</kbd>.
>
> To handle these signals in Rust you need to consider how you can receive these signals as well as how you can react to them.

See #27 for a lot of discussion around signals and crates for handling them!

### PR DESCRIPTION
**Pull request by killercup**
Closes #50
cc #27

### CODE REVIEW

**Comment by killercup** [comment]
r? @stjepang

I tried to keep it simple for now, and hope it's still correct. What more do I need to mention to make this really useful?

File: src/in-depth/signals.md [hunk 1]
```diff
@@ -1 +1,125 @@
...
+you should have a look at [signal-hook].
+Its design is described in [this blog post][signal-hook-post],
+and it is currently the library with the widest community support.
+
+[signal-hook-post]: https://vorner.github.io/2018/06/28/signal-hook.html
+
+## Using channels
+
+Another approach is to use channels:
+You create a channel that the signal handler emits a value on
```
Comment (stjepang) [inline src/in-depth/signals.md]: Perhaps "into which" would read better than "that"?

File: src/in-depth/signals.md [hunk 2]
```diff
@@ -1 +1,125 @@
...
+(a boolean that shareable between threads)
+in your signal handler,
+and in hot loops,
+or when waiting for a thread,
+you periodically check its value
+and break when it becomes true.
+
+## Handling other types of signals
+
+The [ctrlc] crate only handels <kbd>Ctrl</kbd>+<kbd>C</kbd>,
```
Comment (stjepang) [inline src/in-depth/signals.md]: "handels" -> "handles"

File: src/in-depth/signals.md [hunk 3]
```diff
@@ -1 +1,125 @@
...
+[`.into_async()`]: https://docs.rs/signal-hook/0.1.6/signal_hook/iterator/struct.Signals.html#method.into_async
+
+## What to do when you receive another Ctrl+C while you're handling the first Ctrl+C
+
+Most user will press <kbd>Ctrl</kbd>+<kbd>C</kbd>,
+and then give your program a few seconds to exit,
+or tell them what's going on.
+If that doesn't happen,
+they will press <kbd>Ctrl</kbd>+<kbd>C</kbd> again.
+The expected behavior is to have the application quit immediately.
```
Comment (stjepang) [inline src/in-depth/signals.md]: Nit: Maybe "usual" or "typical" rather than "expected"?

File: src/in-depth/signals.md [hunk 4]
```diff
@@ -1 +1,125 @@
...
+
+This is, of course, not that helpful:
+It only prints a message but otherwise doesn't stop the program.
+
+In a real-world program,
+it's a good idea to instead set a variable in the signal handler
+that you then check in various places in your program.
+For example,
+you can set an `Arc<AtomicBool>`
+(a boolean that shareable between threads)
```
Comment (stjepang) [inline src/in-depth/signals.md]: "that is shareable"? Or you could reword this as "a boolean shared between threads".

File: src/in-depth/signals.md [hunk 5]
```diff
@@ -1 +1,125 @@
...
+<aside>
+
+**Note:**
+If your applications does not require special cleanup,
+the default handling is fine
+(i.e. exit immediately and let the OS cleanup).
+In that case:
+No need to do what this chapter tells you!
+For applications that need to clean up after themselves,
+for example to correct close network connections,
```
Comment (stjepang) [inline src/in-depth/signals.md]: "to correctly close"? Or even just "to close".

File: src/in-depth/signals.md [hunk 6]
```diff
@@ -1 +1,125 @@
...
+or reset system settings,
+this chapter is however very relevant!
+
+</aside>
+
+## Differences between operating systems
+
+On Unix systems
+(like Linux, macOS, and FreeBSD)
+a process can receive [signals]
```
Comment (stjepang) [inline src/in-depth/signals.md]: Missing full stop after `[signals]`.

File: src/in-depth/signals-channels.rs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    let (sender, receiver) = bounded(100);
+    ctrlc::set_handler(move || {
+        let _ = sender.send(());
+    })?;
+
+    Ok(receiver)
+}
+
+fn main() -> Result<(), exitfailure::ExitFailure> {
+    let ctrls = ctrl_channel()?;
```
Comment (stjepang) [inline src/in-depth/signals-channels.rs]: Nit: maybe this could be renamed to "let ctrlcs = ctrlc_channel()?;"
But I do see "ctrlcs" as kinda ugly. :/

File: src/in-depth/signals.md [hunk 8]
```diff
@@ -1 +1,125 @@
...
+It can either react to them
+in a default (OS-provided) way,
+catch the signal and handle them in a program-defined way,
+or ignore the signal entirely.
+
+[signals]: https://manpages.ubuntu.com/manpages/bionic/en/man7/signal.7.html
+[signal-safety man page]: http://manpages.ubuntu.com/manpages/bionic/man7/signal-safety.7.html
```
Comment (Screwtapello) [inline src/in-depth/signals.md]: This link is unused?

File: src/in-depth/signals.md [hunk 9]
```diff
@@ -1 +1,125 @@
...
+and break when it becomes true.
+
+## Handling other types of signals
+
+The [ctrlc] crate only handels <kbd>Ctrl</kbd>+<kbd>C</kbd>,
+or, what on Unix systems would be called SIGINT (the "interrupt" signal).
+To react to more Unix signals,
+you should have a look at [signal-hook].
+Its design is described in [this blog post][signal-hook-post],
+and it is currently the library with the widest community support.
```
Comment (stjepang) [inline src/in-depth/signals.md]: Perhaps we could include a simple example with `signal-hook`, something along the lines of:

```rust
fn main() -> Result<(), Box<Error>> {
    let signals = Signals::new(&[SIGINT])?;

    thread::spawn(move || {
        for sig in signals.forever() {
            println!("Received signal {:?}", sig);
        }
    });

    Ok(())
}
```

I didn't try to compile this, though.

File: src/in-depth/signals.md [hunk 10]
```diff
@@ -1 +1,125 @@
...
+
+[signal-hook-post]: https://vorner.github.io/2018/06/28/signal-hook.html
+
+## Using channels
+
+Another approach is to use channels:
+You create a channel that the signal handler emits a value on
+whenever the signal is received.
+In you application code you use
```
Comment (Screwtapello) [inline src/in-depth/signals.md]: "In you" → "In your"

File: src/in-depth/signals.md (hunk 2 above)
Comment (Screwtapello) [inline src/in-depth/signals.md]: "handels" → "handles"

**Comment by stjepang** [comment]
@vorner Just pinging you in case you'd like to take a look.

File: src/in-depth/signals.md [hunk 11]
```diff
@@ -1 +1,125 @@
...
+To react to more Unix signals,
+you should have a look at [signal-hook].
+Its design is described in [this blog post][signal-hook-post],
+and it is currently the library with the widest community support.
+
+[signal-hook-post]: https://vorner.github.io/2018/06/28/signal-hook.html
+
+## Using channels
+
+Another approach is to use channels:
```
Comment (Screwtapello) [inline src/in-depth/signals.md]: Since we've just talked about "handling other types of signals", this sounds like it will be an alternative to the `signal-hook` crate. How about:

> Instead of setting a variable
> and having other parts of the program check it,
> you can use channels:

File: src/in-depth/signals.md [hunk 12]
```diff
@@ -1 +1,125 @@
...
+on signal-hook's `Signals` types
+to get a new type that implements `futures::Stream`.
+
+[signal-hook]: https://crates.io/crates/signal-hook
+[tokio]: https://tokio.rs/
+[`.into_async()`]: https://docs.rs/signal-hook/0.1.6/signal_hook/iterator/struct.Signals.html#method.into_async
+
+## What to do when you receive another Ctrl+C while you're handling the first Ctrl+C
+
+Most user will press <kbd>Ctrl</kbd>+<kbd>C</kbd>,
```
Comment (Screwtapello) [inline src/in-depth/signals.md]: "user" → "users"

File: src/in-depth/signals-channels.rs (hunk 7 above)
Comment (killercup) [reply src/in-depth/signals-channels.rs]: Ah, I meant to write `ctrlcs` actually. My fingers probably found it to weird to type, too!

Would calling it `sig_ints` be better or even more confusing?

Comment (stjepang) [reply src/in-depth/signals-channels.rs]: More confusing, IMO :)

File: src/in-depth/signals.md (hunk 9 above)
Comment (killercup) [reply src/in-depth/signals.md]: Good minimal example. I'll add it!

File: src/in-depth/signals.md (hunk 8 above)
Comment (killercup) [reply src/in-depth/signals.md]: Yeah, I meant to link to it but didn't find a good place, as it's a non-problem if you use a crate calls you hook outside of the signal handler.

File: src/in-depth/signals-channels.rs (hunk 7 above)
Comment (killercup) [reply src/in-depth/signals-channels.rs]: How about `ctrl_c_events`?

**Comment by killercup** [comment]
Thanks so much for the super fast reviews, @Screwtapello and @stjepang! <3

File: src/in-depth/signals-channels.rs (hunk 7 above)
Comment (stjepang) [reply src/in-depth/signals-channels.rs]: Much better, I like it 👍

**Comment by killercup** [comment]
Let's land it! As always, comments here and on #58 are welcome even after the merge!

File: src/in-depth/signals-channels.rs [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+    Ok(receiver)
+}
+
+fn main() -> Result<(), exitfailure::ExitFailure> {
+    let ctrl_c_events = ctrl_channel()?;
+    let ticks = tick(Duration::from_secs(1));
+
+    loop {
+        select! {
```
Comment (jens1o) [inline src/in-depth/signals-channels.rs]: `select!` is deprecated, isn't it?

Comment (killercup) [reply src/in-depth/signals-channels.rs]: This is the select macro from crossbeam-channels, which I think is still going strong

Comment (stjepang) [reply src/in-depth/signals-channels.rs]: This is `crossbeam_channel::select!`, not `std::select!` (which *is* deprecated).

File: src/in-depth/signals.md [hunk 14]
```diff
@@ -1 +1,134 @@
...
+If your applications does not require special cleanup,
+the default handling is fine
+(i.e. exit immediately and let the OS cleanup).
+In that case:
+No need to do what this chapter tells you!
+For applications that need to clean up after themselves,
+for example to close network connections,
+remove temporary files,
+or reset system settings,
+this chapter is however very relevant!
```
[the remark this answers is not in the document]
Comment (vorner) [reply src/in-depth/signals.md]: The other chapter probably sounds better (and it probably is not specific to CLI only). Still, as a wording nitpick, maybe not call it „cleanup“, but „graceful shutdown“? Like, the OS will terminate network connections, but it might happen in the middle of a message instead of after saying proper „Good bye“.

[the remark this answers is not in the document]
Comment (killercup) [reply src/in-depth/signals.md]: Thanks! Expanded the note to describe graceful shutdown scenarios, and opened #107 for tracking ideas around a "robustness" chapter.

File: src/in-depth/signals-channels.rs (hunk 13 above)
Comment (jens1o) [reply src/in-depth/signals-channels.rs]: ah, pardon me. :)
